app/models/bot/tagger.rb

```diff
@@ -0,0 +1,65 @@
+require 'json'
+
+class Bot::Tagger < BotUser
+  check_settings
+  class Error < ::StandardError
+  end
+
+  def self.get_tag_text(tag_id,auto_tag_prefix,ignore_autotags)
+    tag=TagText.find_by_id(tag_id).text
+    if tag.nil? || (ignore_autotags && tag[0]==auto_tag_prefix)
+      return nil
+    else
+      tag[0]==auto_tag_prefix ? tag[1..] : tag
+    end
+  end
+
+  def self.run(body)
+    Rails.logger.info("[AutoTagger Bot] Received event with body of #{body}")
+    if CheckConfig.get('alegre_host').blank?
+      Rails.logger.warn("[AutoTagger Bot] Skipping events because `alegre_host` config is blank")
+      return false
+    end
+
+    handled = false
+    pm = nil
+    begin
+      settings=JSON.parse(body["settings"])
+      auto_tag_prefix=settings["auto_tag_prefix"]
+      threshold=settings["threshold"]/100.0
+      ignore_autotags=settings["ignore_autotags"]
+      pm = ProjectMedia.where(id: body.dig(:data, :dbid)).last
+      if body.dig(:event) == 'create_project_media' && !pm.nil?
+        Rails.logger.info("[AutoTagger Bot] [ProjectMedia ##{pm.id}] This item was just created, processing...")
+        results=Bot::Alegre.get_items_with_similar_text(pm, Bot::Alegre::ALL_TEXT_SIMILARITY_FIELDS,
+          [{ value: threshold }], pm.title, [Bot::Alegre.matching_model_to_use(pm.team_id)].flatten.reject{|m| m==Bot::Alegre::ELASTICSEARCH_MODEL})
+        Rails.logger.info("[AutoTagger Bot] [ProjectMedia ##{pm.id}] #{results.length} nearest neighbors #{results.keys()}")
+        Rails.logger.info("[AutoTagger Bot] [ProjectMedia ##{pm.id}] Results: #{results}")
+        tag_counts=results.map{|nn_pm,_| ProjectMedia.find(nn_pm).get_annotations('tag')}.flatten.map{|t| self.get_tag_text(t[:data][:tag],auto_tag_prefix,ignore_autotags)}.group_by(&:itself).transform_values(&:count)
+        tag_counts=tag_counts.reject{|t|t==nil}.sort_by{|_k,v| v}
+        Rails.logger.info("[AutoTagger Bot] [ProjectMedia ##{pm.id}] Tag distribution #{tag_counts}")
+        if tag_counts.length > 0
+          max_count=tag_counts.last[1]
```

Comment:
**question** I'm a little confused by this - is tag_counts an array of arrays? if it's not (which I think it isn't, based on line 39) then `tag_counts.last` should give you the last item of the array, and `tag_counts.last[1]` might give you nil since you're accessing a non-existent index

Reply:
`tag_counts` is an array of arrays. For example, `[['nature', 1], ['sport', 2]]` . Thus `.last[1]` gives the count of the most numerous tag.  I agree the reject of line 39 seems strange in that case. I'll double check.